File: android/nativepreview/src/main/java/com/travelplaner/nativepreview/domain/MemoryBackup.kt
package com.travelplaner.nativepreview.domain

import kotlinx.serialization.json.*

object MemoryBackup {
    fun encode(trip: TripDocument, imageResolver: (String) -> ByteArray?): ByteArray {
        val root = buildJsonObject {
            trip.json.forEach { (key, value) -> if (key != "sharedId" && key != "sharedManagementToken") put(key, value) }
            val entries = trip.json["journalEntries"]?.jsonArray
            if (entries != null) {
                put("journalEntries", JsonArray(entries.map { element ->
                    val entry = element.jsonObject.toMutableMap()
                    val fileName = entry["imageFileName"]?.jsonPrimitive?.contentOrNull
                    if (fileName != null) {
                        val bytes = imageResolver(fileName) ?: throw IllegalArgumentException("기록 사진을 찾지 못했어요.")
                        entry["imageDataUrl"] = JsonPrimitive(dataUrl(bytes))
                        entry.remove("imageFileName")
                    }
                    JsonObject(entry)
                }))
            }
        }
        val bytes = root.toString().encodeToByteArray()
        require(bytes.size <= TripBackup.maxBytes) { "백업은 최대 20 MiB까지 내보낼 수 있어요." }
        return bytes
    }

    private fun dataUrl(bytes: ByteArray): String {
        val png = bytes.size >= 8 && bytes.copyOfRange(0, 8).contentEquals(byteArrayOf(0x89.toByte(),0x50,0x4e,0x47,0x0d,0x0a,0x1a,0x0a))
        return "data:${if (png) "image/png" else "image/jpeg"};base64,${java.util.Base64.getEncoder().encodeToString(bytes)}"
    }
}

File: ios/NativeCore/Sources/NativeCore/MemoryBackup.swift
import Foundation

public enum MemoryBackup {
    public static func encode(_ trip: TripDocument, imageResolver: (String) throws -> Data?) throws -> Data {
        var raw = trip.raw
        raw.removeValue(forKey: "sharedId")
        raw.removeValue(forKey: "sharedManagementToken")
        if let entries = raw["journalEntries"] as? [[String: Any]] {
            raw["journalEntries"] = try entries.map { entry in
                var copy = entry
                if let fileName = copy["imageFileName"] as? String {
                    guard let data = try imageResolver(fileName) else { throw BackupError.invalid("기록 사진을 찾지 못했습니다.") }
                    copy["imageDataUrl"] = dataURL(data)
                    copy.removeValue(forKey: "imageFileName")
                }
                return copy
            }
        }
        let data = try JSONSerialization.data(withJSONObject: raw, options: [.sortedKeys, .prettyPrinted])
        guard data.count <= TripBackup.maxBytes else { throw BackupError.tooLarge }
        return data
    }

    private static func dataURL(_ data: Data) -> String {
        let mime = data.starts(with: [0x89,0x50,0x4e,0x47,0x0d,0x0a,0x1a,0x0a]) ? "image/png" : "image/jpeg"
        return "data:\(mime);base64,\(data.base64EncodedString())"
    }
}
